src/node.rs

```diff
@@ -445,6 +445,84 @@ impl PbftNode {
         Ok(())
     }
 
+    /// Attempts to catch this node up with its peers, if necessary
+    ///
+    /// Returns true if catching up was required, and false otherwise
+    fn try_catchup(
+        &mut self,
+        state: &mut PbftState,
+        block: &Block,
+        msg: PbftMessage,
+    ) -> Result<bool, PbftError> {
+        if state.is_primary() {
+            return Ok(false);
+        }
+
+        // If we've got a (Tentative)WorkingBlock, and the new block is immediately
+        // subsequent to it, then we're able to catch up. Otherwise, return false
+        // to signify that no catching up occurred.
+        match state.working_block.clone() {
+            WorkingBlockOption::WorkingBlock(wb) => {
+                let block_num_matches = block.block_num == wb.get_block_num() + 1;
+                let block_id_matches = block.previous_id == wb.get_block_id();
+
+                if !block_num_matches || !block_id_matches {
+                    return Ok(false);
+                }
+            }
+            WorkingBlockOption::TentativeWorkingBlock(bid) => {
+                if block.previous_id == bid {
+                    // If we've got a tentative working block, replace it with a regular working block
+                    state.working_block = WorkingBlockOption::WorkingBlock(msg.get_block().clone());
+                } else {
+                    return Ok(false);
```

Comment:
It is not clear to me why we don't want to do catchup when we have a TentativeWorkingBlock. From what I can tell, this just means we haven't received any pre-prepare messages from the rest of the network about this block. Receiving a valid block built on the tentative working block would indicate that those messages do in fact exist and that we just haven't received them yet.

Reply:
Lol, nevermind I just realized I was misreading this.